### PR TITLE
[TS-682] Fixed encoding and decoding virtual messages collections and their fields.

### PR DESCRIPTION
**Pull request by Ilya-Loginov-Exactpro**

### CODE REVIEW

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt [hunk 1]
```diff
@@ -371,6 +364,32 @@ open class XmlPipelineCodec : IPipelineCodec {
...
+                }
+
+                message[fieldName] = listValue
+            } else {
+                val virtualMessage = message()
+                decodeVirtualMessage(fields, virtualMessage, fieldStructure)
+                if (fields.isNotEmpty()) {
+                    throw DecodeException("Can not decode field with name '$fieldName'. Not all field decoded for virtual message with name '${fieldStructure.name}'")
+                }
+                message[fieldName] = virtualMessage
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt]: Let's implement the logic
1) create a message
2) check does the current message contain the current node
2-yes.1) check if the current message structure collection
2-yes.1-yes) store current message, create new, go to 2-no.1
2-yes.1-no) throw the exception "... the node has duplicated field ..."
2-no.1) put field, go to the next initiation

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt [hunk 2]
```diff
@@ -47,10 +52,8 @@ class XmlMessageStructureVisitor(private val document: Document, private val nod
             } else {
                 val nodeName = fldStruct.getXmlTagName()
                 val valueNode = node.findNode(nodeName)
-                if (valueNode == null) {
+                if (valueNode == null || valueNode.getText() != fieldValue) {
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt]: Maybe add else for this case

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt [hunk 3]
```diff
@@ -125,28 +124,14 @@ class XmlMessageStructureVisitor(
...
-                    ),
-                    message,
-                    isCollection
-                ), fldStruct
-            )
-            return
-        }
+        val newNode = if (fldStruct.isVirtual()) node else
+            if (isCollection || multiplyParent) node.addNode(xmlTagName, document)
+            else node.findOrAddNode(xmlTagName, document)
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt]:
```        
val newNode = when {
    fldStruct.isVirtual() -> node
    isCollection || multiplyParent -> node.addNode(xmlTagName, document)
    else -> node.findOrAddNode(xmlTagName, document)
}
```

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt (hunk 2 above)
Comment (Ilya-Loginov-Exactpro) [reply src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt]: Why? In this method, we encode a single string field. Fields in the virtual collection messages need this condition. If we found a node with an equals value we should not create a new node. We can change `valueNode.getText() != fieldValue` to `multiplyParent`.

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Pull request merged**